### PR TITLE
[ie/generic] Add `impersonate` extractor arg

### LINKED ISSUE #11150: Memedroid
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

Brazil, EUA, French, Italy, Spain, Portugal and others.

### Example URLs

- Single Video: https://www.memedroid.com/memes/detail/4373699/OMW-No-Homo

### Provide a description that is worded well enough to be understood

To extract the video from the website it seems easy, just search for the `source` with the name `detailed-item-video-src-mp4` in the html:

```html
<source id="detailed-item-video-src-mp4" src="https://rvideos1.memedroid.com/videos/UPLOADED920/66fdc31a31e66.mp4" type="video/mp4">
```

- Source file in the HTML: https://rvideos2.memedroid.com/videos/UPLOADED854/66fd685e93c5b.mp4

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
🞈 yt-dlp -vU https://www.memedroid.com/memes/detail/4373699/OMW-No-Homo
[debug] Command-line config: ['-vU', 'https://www.memedroid.com/memes/detail/4373699/OMW-No-Homo']
[debug] User config "/var/home/heus/.config/yt-dlp/config": ['--ignore-errors', '-o', '~/Videos/%(title)s.%(ext)s', '-f', 'bestvideo[ext=mp4][width<2000][height<=1200]+bestaudio[ext=m4a]/bestvideo[ext=webm][width<2000][height<=1200]+bestaudio[ext=webm]/bestvideo[width<2000][height<=1200]+bestaudio/best[width<2000][height<=1200]/best']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2024.09.27 from yt-dlp/yt-dlp [c6387abc1] (pip)
[debug] Python 3.12.6 (CPython x86_64 64bit) - Linux-6.10.11-200.fc40.x86_64-x86_64-with-glibc2.39 (OpenSSL 3.0.14 4 Jun 2024, glibc 2.39)
[debug] exe versions: ffmpeg 6.1.2 (setts), ffprobe 6.1.2
[debug] Optional libraries: Cryptodome-3.21.0, brotli-1.1.0, certifi-2024.08.30, mutagen-1.47.0, requests-2.32.3, sqlite3-3.46.0, urllib3-2.2.3, websockets-13.1
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Loaded 1838 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2024.09.27 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2024.09.27 from yt-dlp/yt-dlp)
[generic] Extracting URL: https://www.memedroid.com/memes/detail/4373699/OMW-No-Homo
[generic] OMW-No-Homo: Downloading webpage
ERROR: [generic] Unable to download webpage: HTTP Error 403: Forbidden (caused by <HTTPError 403: Forbidden>)
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 741, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/extractor/generic.py", line 2384, in _real_extract
    full_response = self._request_webpage(url, video_id, headers=filter_dict({
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 910, in _request_webpage
    raise ExtractorError(errmsg, cause=err)

  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 897, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query, extensions))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/YoutubeDL.py", line 4165, in urlopen
    return self._request_director.send(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/networking/common.py", line 117, in send
    response = handler.send(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/networking/_helper.py", line 208, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/networking/common.py", line 340, in send
    return self._send(request)
           ^^^^^^^^^^^^^^^^^^^
  File "/var/home/heus/.local/pipx/venvs/yt-dlp/lib/python3.12/site-packages/yt_dlp/networking/_requests.py", line 365, in _send
    raise HTTPError(res, redirect_loop=max_redirects_exceeded)
yt_dlp.networking.exceptions.HTTPError: HTTP Error 403: Forbidden
```

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Adds the `impersonate` extractor arg to the generic extractor.
Also automatically tries to use impersonation for the generic website request if impersonation is available.

Fixes #11150

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
</details>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1795,6 +1795,7 @@ The following extractors use this feature:
 * `key_query`: Passthrough the master m3u8 URL query to its HLS AES-128 decryption key URI if no value is provided, or else apply the query string given as `key_query=VALUE`. Note that this will have no effect if the key URI is provided via the `hls_key` extractor-arg. Does not apply to ffmpeg
 * `hls_key`: An HLS AES-128 key URI *or* key (as hex), and optionally the IV (as hex), in the form of `(URI|KEY)[,IV]`; e.g. `generic:hls_key=ABCDEF1234567980,0xFEDCBA0987654321`. Passing any of these values will force usage of the native HLS downloader and override the corresponding values found in the m3u8 playlist
 * `is_live`: Bypass live HLS detection and manually set `live_status` - a value of `false` will set `not_live`, any other value (or no value) will set `is_live`
+* `impersonate`: Targets to try and impersonate in the website request; e.g. `safari,chrome-110`. By default any available target will be used. Set to `false` to not use impersonation.
```
Comment (gamer191) [inline README.md:1798]: This does the exact same thing as `--impersonate`, except that it's enabled by default

I think a better solution (from a UI perspective, I don't know about the code perspective) would be to just add a --no-impersonate option

Comment (Grub4K) [reply README.md:1798]: No it does not. Its only for the website request. Having `--no-impersonate` would be obsoleted by `--impersonate false` once impersonate handlers will support multiple targets

Comment (bashonly) [reply README.md:1798]:
```suggestion
* `impersonate`: Target(s) to try and impersonate with the initial webpage request; e.g. `safari,chrome-110`. By default any available target will be used. Use `false` to disable impersonation
```

File: yt_dlp/extractor/generic.py [hunk 2]
```diff
@@ -2373,6 +2374,14 @@ def _real_extract(self, url):
         else:
             video_id = self._generic_id(url)
 
+        # Try to impersonate a web-browser by default if possible
+        # We test if impersonation is available at all to omit the warning if it isn't
+        impersonate = None
+        if self._downloader._impersonate_target_available(ImpersonateTarget()):
+            targets = self._configuration_arg('impersonate', [''])
+            if targets != ['false']:
+                impersonate = list(map(ImpersonateTarget.from_str, targets))
```
Comment (bashonly) [inline yt_dlp/extractor/generic.py:2383]: this kinda duplicates what we are already doing in `InfoExtractor._request_webpage`. we could simplify it like this:
```suggestion
        impersonate = self._configuration_arg('impersonate', [''])
        if 'false' in impersonate or not self._downloader._impersonate_target_available(ImpersonateTarget()):
            impersonate = None
```

**Review by bashonly — APPROVED** [review]

**Pull request merged**